src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/BlazorDebug/BlazorDebugConfigurationProvider.ts

```diff
@@ -13,29 +13,69 @@ export class BlazorDebugConfigurationProvider implements vscode.DebugConfigurati
     constructor(private readonly logger: RazorLogger, private readonly vscodeType: typeof vscode) { }
 
     public async resolveDebugConfiguration(folder: vscode.WorkspaceFolder | undefined, configuration: vscode.DebugConfiguration): Promise<vscode.DebugConfiguration | undefined> {
-        const shellPath = process.platform === 'win32' ? 'cmd.exe' : 'dotnet';
-        const shellArgs = process.platform === 'win32' ? ['/c', 'chcp 65001 >NUL & dotnet run'] : ['run'];
-        const spawnOptions = {
-            cwd: configuration.cwd || (folder && folder.uri && folder.uri.fsPath),
-        };
-
-        const output = this.vscodeType.window.createTerminal({
-          name: 'Blazor WebAssembly App',
-          shellPath,
-          shellArgs,
-          ...spawnOptions,
-        });
-
         /**
-         * On non-Windows platforms, we need to terminate the Blazor
-         * dev server and its child processes.
+         * If the user is debugging a hosted Blazor WebAssembly application,
+         * then the application server needs to be launched via VS Code's debug
+         * workflow to allow debugging on the server-side logic. If the app is
+         * not hosted, then it can be spun as a standalone process to avoid additional overhead.
          */
-        const terminate = this.vscodeType.debug.onDidTerminateDebugSession(async event => {
-            await onDidTerminateDebugSession(event, this.logger, await output.processId);
-            terminate.dispose();
-        });
+        if (configuration.hosted) {
+            if (!configuration.program) {
+                const message = `Must provide 'program' property in launch configuration for hosted Blazor WebAssembly apps.`;
+                this.vscodeType.window.showErrorMessage(message);
+            }
+
+            const app = {
+                name: '.NET Core Launch (Blazor Hosted)',
+                type: 'coreclr',
+                request: 'launch',
+                preLaunchTask: 'build',
+                program: configuration.program,
+                args: [],
+                // Run the `Server` project in hosted apps
+                cwd: '${workspaceFolder}/Server',

```

Comment:
Can't assume /Server either 😢

Reply:
Bleh, you are correct. We'll have to read this from the configuration as well.

In the future, I'd like to set up a prompt UI for these fields so the user experience doesn't involve manually editing JSON.

Reply:
Fixed.

Reply:
> In the future, I'd like to set up a prompt UI for these fields so the user experience doesn't involve manually editing JSON.

You can totally do this as part of O#'s launch.json generation if you want 😄